src/java.base/share/classes/java/lang/Long.java

```diff
@@ -446,33 +488,105 @@ private static void formatUnsignedLong0UTF16(long val, int shift, byte[] buf, in
+        char i14 = hex256[(((int) lsb) >> 8) & 0xff];
+        char i15 = hex256[((int) lsb) & 0xff];
+
         if (COMPACT_STRINGS) {
             byte[] buf = new byte[36];
-            formatUnsignedLong0(lsb,        4, buf, 24, 12);
-            formatUnsignedLong0(lsb >>> 48, 4, buf, 19, 4);
-            formatUnsignedLong0(msb,        4, buf, 14, 4);
-            formatUnsignedLong0(msb >>> 16, 4, buf, 9,  4);
-            formatUnsignedLong0(msb >>> 32, 4, buf, 0,  8);
 
-            buf[23] = '-';
-            buf[18] = '-';
+            buf[0] = (byte) (i0 >> 8);
+            buf[1] = (byte) i0;
+            buf[2] = (byte) (i1 >> 8);
+            buf[3] = (byte) i1;
+            buf[4] = (byte) (i2 >> 8);
+            buf[5] = (byte) i2;
+            buf[6] = (byte) (i3 >> 8);
+            buf[7] = (byte) i3;
+            buf[8] = '-';
+            buf[9] = (byte) (i4 >> 8);
+            buf[10] = (byte) i4;
+            buf[11] = (byte) (i5 >> 8);
+            buf[12] = (byte) i5;
             buf[13] = '-';
-            buf[8]  = '-';
+            buf[14] = (byte) (i6 >> 8);
+            buf[15] = (byte) i6;
+            buf[16] = (byte) (i7 >> 8);
+            buf[17] = (byte) i7;
+            buf[18] = '-';
+            buf[19] = (byte) (i8 >> 8);
+            buf[20] = (byte) i8;
+            buf[21] = (byte) (i9 >> 8);
+            buf[22] = (byte) i9;
+            buf[23] = '-';
+            buf[24] = (byte) (i10 >> 8);
+            buf[25] = (byte) i10;
+            buf[26] = (byte) (i11 >> 8);
+            buf[27] = (byte) i11;
+            buf[28] = (byte) (i12 >> 8);
+            buf[29] = (byte) i12;
+            buf[30] = (byte) (i13 >> 8);
+            buf[31] = (byte) i13;
+            buf[32] = (byte) (i14 >> 8);
+            buf[33] = (byte) i14;
+            buf[34] = (byte) (i15 >> 8);
+            buf[35] = (byte) i15;
 
             return new String(buf, LATIN1);
         } else {
             byte[] buf = new byte[72];
 
-            formatUnsignedLong0UTF16(lsb,        4, buf, 24, 12);
-            formatUnsignedLong0UTF16(lsb >>> 48, 4, buf, 19, 4);
-            formatUnsignedLong0UTF16(msb,        4, buf, 14, 4);
-            formatUnsignedLong0UTF16(msb >>> 16, 4, buf, 9,  4);
-            formatUnsignedLong0UTF16(msb >>> 32, 4, buf, 0,  8);
-
-            StringUTF16.putChar(buf, 23, '-');
-            StringUTF16.putChar(buf, 18, '-');
+            StringUTF16.putChar(buf, 0, (byte) (i0 >> 8));
+            StringUTF16.putChar(buf, 1, (byte) i0);
+            StringUTF16.putChar(buf, 2, (byte) (i1 >> 8));
+            StringUTF16.putChar(buf, 3, (byte) i1);
+            StringUTF16.putChar(buf, 4, (byte) (i2 >> 8));
+            StringUTF16.putChar(buf, 5, (byte) i2);
+            StringUTF16.putChar(buf, 6, (byte) (i3 >> 8));
+            StringUTF16.putChar(buf, 7, (byte) i3);
+            StringUTF16.putChar(buf, 8, '-');
+            StringUTF16.putChar(buf, 9, (byte) (i4 >> 8));
```

Comment:
This might be cheating but you could avoid a store of 0 to the high byte (its already zero) by inlining the code from StringUTF16.putChar();  just double the index on the store of the byte.
```
            buf[0] = (byte) (i0 >> 8);
            buf[2] = (byte) i0;
            buf[4] = (byte) (i1 >> 8);
            buf[6] = (byte) i1;
            buf[8] = (byte) (i2 >> 8);
...
```

Reply:
Note that `StringUTF16::putChar` takes endian‑ness into account, so the above code would only be correct on half of the supported endian types.

Reply:
i have simplified the code using ENDIAN and COMPACT_STRINGS